<a href="https://colab.research.google.com/github/jmrosarosa/KSchoolMDS/blob/main/Copy_of_%5BKshool_NB003%5DPython_Variables.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [Kshool NB003]Python_Variables
* Objetivos: 
    - Dotar al alumno de conocimiento en el entendimiento y trabajo con las variables simple de Int, Floats y Bool
    - Comprender y profundizar en el tipo complejo de dato String y sus métodos.

<img src="https://media.istockphoto.com/id/1024173328/vector/challenge-accepted-banner.jpg?s=612x612&w=0&k=20&c=HkpbXPIxSoS0zZIB36AdXz8u3TE2peAC0_s1jNTCKTc=" alt="Challenge Accepted" style="width: 300px">

---
---
---

# **Tipos de Datos**
---

`Python` tiene varios tipos de datos, los mismos datos básicos que otros lenguajes:
- enteros
- flotantes
- strings
- booleanos. 

Además, las **listas** son un tipo predefinido en el lenguaje, los **diccionarios**, **tuplas** y **sets**.

La funcion que se utiliza para determinar el tipo de datos es `type()`

In [ ]:
#Funcion para determinar el tipo de dato numerico

print(type(3)) # tipo entero (int)
print(type(3.0)) # Tipo flotante, se determina por el uso del punto (float)
print(type(2+2j)) # Numero complejo
print(type(5/2)) # Numero flotante (float)
print(type(5//2)) # Numero entero (int)

<class 'int'>
<class 'float'>
<class 'complex'>
<class 'float'>
<class 'int'>


# **Variables**
---

Siempre que creas una **variable** en **Python**, tiene un valor con su correspondiente tipo de datos. 

Hay muchos tipos de datos diferentes, como enteros, flotantes, booleanos y cadenas, que trataremos en esta lección.

> *Una Variable es como una caja en la memoria del ordenador donde puedes almacenar un único valor. - Al Sweigart*

A diferencia de otros muchos lenguajes de programación, en Python **NO** necesitas **declarar** ninguna variable ni **
**.
El nombre_de_la_variable en Python puede ser tan corto como dulce como a, b, x, y, ... o puede ser muy informativo como edad, altura, nombre, nombre_del_alumno, covid, ...

> *Aunque se recomienda mantener un nombre de variable muy descriptivo para mejorar la legibilidad.*


**Reglas**
- Una vez hecho esto, hay algunas reglas que debes seguir para nombrar una variable:
    - El nombre de una variable debe empezar por una *letra* o el *carácter de subrayado*.
    - El nombre de una variable no puede empezar por un *número*.
    - Un nombre de variable sólo puede contener *caracteres alfanuméricos* y *guiones bajos*. Por ejemplo, algo como esto es válido A-z, 0-9 y _*
    - Los nombres de las variables distinguen entre *mayúsculas y minúsculas* *(altura, Altura y ALTURA* son tres nombres de variables diferentes)

Los tipos de datos son importantes, porque determinan qué tipo de acciones puedes hacer con ellos. 

Por ejemplo, puedes dividir dos flotantes, pero no puedes dividir dos cadenas. Por ejemplo, `12.0/2.0` tiene sentido, pero `gato`/`perro` no.

Para evitar errores, tenemos que asegurarnos de que las acciones coinciden con los tipos de datos que tenemos.

**Nombres de variables válidos:**
- firstname
- lastname
- num1

**Nombres de variables no válidos:**
- first-name
- first@name
- 1num

La siguiente celda de código crea una variable llamada `test_var` y le asigna el valor que obtenemos al sumar 5 a 4.

A continuación imprimimos el valor asignado a la variable, que es **9**.

In [ ]:
# Crea una variable llamada test_var y dale un valor de 4+5
test_var = 4 + 5

# Imprime el valor de test_var
print(test_var)

9


En general, para trabajar con una variable, tienes que empezar por seleccionar el nombre que quieres utilizar. 

- Lo ideal es que los nombres de las variables sean cortos y descriptivos. También deben cumplir varios requisitos:
    - No pueden tener espacios (por ejemplo, test var no está permitido)
    - Sólo pueden incluir letras, números y guiones bajos (por ejemplo, ¡var_prueba! no está permitido)
    - Tienen que empezar por una letra o un guión bajo (por ejemplo, 1_var no está permitido)

Luego, para crear la variable, tienes que utilizar `=` para asignarle el valor que quieres que tenga.

Siempre puedes echar un vistazo al valor asignado a la variable utilizando `print()` y poniendo el nombre de la variable entre paréntesis.

Con el tiempo, aprenderás a seleccionar buenos nombres para las variables de Python. 

> Está completamente bien que ahora te sientas incómodo, ¡y la mejor forma de aprender es viendo un montón de código Python!

# **Números enteros**
---

Los números enteros (`int`) son números sin parte fraccionaria y pueden ser **positivos** (1, 2, 3, ...), **negativos** (-1, -2, -3, ...) o **cero** (0).

En la celda de código siguiente, establecemos una variable **x** como un número entero. A continuación, verificamos el tipo de datos con `type()`, y sólo tenemos que pasar el nombre de la variable entre paréntesis.

In [ ]:
x = 14
print(x)
print(type(x))

14
<class 'int'>


# **Números Flotantes**
---

Los flotantes (`floats`) son números con partes fraccionarias. Pueden tener muchas cifras después del decimal.

In [ ]:
nearly_pi = 3.141592653589793238462643383279502884197169399375105820974944
print(nearly_pi)
print(type(nearly_pi))

3.141592653589793
<class 'float'>


También podemos especificar un flotante con una fracción.

In [ ]:
almost_pi = 22/7
print(almost_pi)
print(type(almost_pi))

3.142857142857143
<class 'float'>


Una función especialmente útil para las fracciones es la función `round()`. Te permite redondear un número a un número determinado de decimales.

In [ ]:
# Redondea a 5 decimales
rounded_pi = round(almost_pi, 5)
print(rounded_pi)
print(type(rounded_pi))

3.14286
<class 'float'>


Siempre que escribas un número con punto decimal, Python lo reconocerá como un tipo de dato **float**.

Por ejemplo, `1.` (o 1,0, 1,00, etc) será reconocido como un flotante. 
Esto es así, ¡aunque estos números técnicamente no tengan parte fraccionaria!

In [ ]:
y_float = 1.
print(y_float)
print(type(y_float))

1.0
<class 'float'>


## Tipos de datos numéricos en Python:
---

- **Números enteros**: Números enteros(negativos, cero y positivos) --> Ejemplo: ... -3, -2, -1, 0, 1, 2, 3 ...
- **Números en coma flotante**(números decimales) --> Ejemplo: ... -3,5, -2,25, -1,0, 0,0, 1,1, 2,2, 3,5 ...
- **Números complejos** --> Ejemplo: 1 + j, 2 + 4j, 1 - 1j

En el código de abajo se representa el valor máximo por tipología de enteros y flotantes. Esto lo veréis en el módulo de **Numpy** pero es bueno que entendáis las diferentes tipologías y sus rangos, ya que entre otras cosas os puede servir ***para optimizar el código en haras de disminuir el tamaño de la información a trabajar***

In [ ]:
import numpy as np

# Tipos de datos enteros
integer_types = [np.int8, np.int16, np.int32, np.int64]

# Tipos de datos de punto flotante
floating_types = [np.float16, np.float32, np.float64]

print("Valores máximos para tipos de datos enteros:")
for int_type in integer_types:
    max_value = np.iinfo(int_type).max
    print(f"{int_type.__name__}: {max_value}")

print("\nValores máximos para tipos de datos de punto flotante:")
for float_type in floating_types:
    max_value = np.finfo(float_type).max
    print(f"{float_type.__name__}: {max_value}")

Valores máximos para tipos de datos enteros:
int8: 127
int16: 32767
int32: 2147483647
int64: 9223372036854775807

Valores máximos para tipos de datos de punto flotante:
float16: 65504.0
float32: 3.4028234663852886e+38
float64: 1.7976931348623157e+308


| Tipo de dato | Valor máximo                     |
|--------------|----------------------------------|
| int8         | 127                              |
| int16        | 32,767                           |
| int32        | 2,147,483,647                    |
| int64        | 9,223,372,036,854,775,807        |
| float16      | 65,504.0                         |
| float32      | 3.4028235 × 10^38                |
| float64      | 1.7976931348623157 × 10^308      |


# **Booleanos**
---

Los booleanos representan uno de dos valores: `True` o `False`. En la celda de código siguiente, `z_one` se establece como un booleano con valor `True`

In [ ]:
z_one = True
print(z_one)
print(type(z_one))

True
<class 'bool'>


A continuación, `z_two` se establece como un booleano con valor `False`.

In [ ]:
z_two = False
print(z_two)
print(type(z_two))

False
<class 'bool'>


Los booleanos se utilizan para representar el valor de verdad de una expresión. Como `1<2` es una afirmación **verdadera**, `z_three` toma el valor `True`.

In [ ]:
z_three = (1 < 2)
print(z_three)
print(type(z_three))

True
<class 'bool'>


Del mismo modo, como `5 < 3` es una afirmación **falsa**, `z_four` toma el valor `False`.

In [ ]:
z_four = (5 < 3)
print(z_four)
print(type(z_four))

False
<class 'bool'>


Podemos cambiar el valor de un booleano utilizando `not`. Así, `not True` equivale a `False`, y `not False` se convierte en `True`.

In [ ]:
z_five = not z_four
print(z_five)
print(type(z_five))

True
<class 'bool'>


# **Strings**
---

El tipo de datos string o cadena, es una colección de caracteres (como letras del alfabeto, signos de puntuación, dígitos numéricos o símbolos) contenidos entre comillas.

Las cadenas se utilizan habitualmente para representar texto.

In [ ]:
w = "¡Hola, Python!"
print(w)
print(type(w))

¡Hola, Python!
<class 'str'>


Puedes obtener la longitud de una cadena con `len()`. "¡Hola, Python!" tiene una longitud de 14, porque tiene 14 caracteres, incluyendo el espacio, la coma y el signo de exclamación. 

Ten en cuenta que las comillas no se incluyen al calcular la longitud.

In [ ]:
print(len(w))

14


Un tipo especial de string es la cadena vacía, que tiene longitud cero.

In [ ]:
cadena_breve = ""
print(type(cadena_breve))
print(len(cadena_breve))

<class 'str'>
0


Si pones un número entre comillas, tiene un tipo de dato cadena.

In [ ]:
numero = "1.12321"
print(numero)
print(type(numero))

1.12321
<class 'str'>


Si tenemos un string convertible a un float, podemos utilizar float().

In [ ]:
numero2 = float(numero)
print(numero2)
print(type(numero2))

1.12321
<class 'float'>


##### **¡Esto no siempre funcionará! Por ejemplo, podemos convertir "10,43430" y "3" en flotantes, pero no podemos convertir "¡Hola, Python!" en un flotante.**

Igual que puedes sumar dos números (flotantes o enteros), también puedes sumar dos **strings**. El resultado es una cadena más larga que combina las dos cadenas originales concatenándolas.

In [ ]:
new_string = "abc" + "def"
print(new_string)
print(type(new_string))

abcdef
<class 'str'>


Ten en cuenta que no es posible hacer una resta o una división con dos cadenas.
Tampoco puedes multiplicar dos cadenas, pero puedes multiplicar una cadena por un número entero. 

Esto también da como resultado una cadena que no es más que la cadena original concatenada consigo misma un número determinado de veces.

In [ ]:
newest_string = "abc" * 3
print(newest_string)
print(type(newest_string))

abcabcabc
<class 'str'>


Ten en cuenta que no puedes multiplicar una cadena por un flotante. Si lo intentas, lo único que vamos a obtener es un **error**

In [ ]:
will_not_work = "abc" * 3.
will_not_work

TypeError: can't multiply sequence by non-int of type 'float'

> Vemos que tenemos un error de tipo "TypeError", este y otros errores los veremos en el módulo final de Python

#### Secuencia de escape utilizando strings
---

En las cadenas de Python, la barra invertida "`\`" es un carácter especial, también llamado carácter de "escape". 
Se utiliza para representar ciertos caracteres de espacio en blanco: 
- \n: nueva línea
- \t: Tabulador medio(8 espacios)
- \r: Retorno de carro
- \': Comilla simple (')
- \": Comilla doble (")

In [ ]:
# Esta es una secuencia de escape.
string = "Esto es un cuaderno de \"Google Colab\" "
print(string)

Esto es un cuaderno de "Google Colab" 


#### Acceder al string por un índice
---

Se puede acceder a las cadenas por su índice para obtener el valor. Para ello, sólo tienes que colocar el número (valor del índice) dentro del par de corchetes junto con el nombre de la cadena.

In [ ]:
string = "Python"
print(string[2])

t


In [ ]:
print(string[5])

n


#### Sliccing una cadena
---

Slice (trozear) una cadena o string ayuda a obtener un conjunto de caracteres de una cadena. 
Esto es muy útil cuando queremos acceder a un conjunto concreto de caracteres de una cadena. Abajo te indico algunas variantes de slice que te pueden resultar útiles.

In [ ]:
string = "Python en Kschool"

print(f"Para la variable string {string}\n")

#Obtener un carácter de la cadena
print(string[0:1])

#Obtener los tres primeros caracteres de la cadena
print(string[0:3])

#Obtener los tres primeros caracteres de la cadena (Alternativa)
print(string[:3])

#Obtener los tres últimos caracteres de la cadena
print(string[-3:])

#Obtener todos los caracteres pero excluyendo los tres primeros caracteres de la cadena
print(string[3:])

#Obtener todos los caracteres pero excluyendo los tres últimos de la cadena
print(string[:-3])

#Invertir todos los caracteres de una cadena dada
print(string[::-1])

#Alternativa para imprimir todos los caracteres de una cadena
print(string[::])

Para la variable string Python en Kschool

P
Pyt
Pyt
ool
hon en Kschool
Python en Ksch
loohcsK ne nohtyP
Python en Kschool


#### Dividir una cadena
---

A veces, dividir una cadena es una opción útil porque es una de las formas más sencillas de convertir una cadena en una lista. Sé que no he hablado de la lista, pero ten en cuenta que dividir convierte una cadena en una lista.

In [ ]:
string = "Python en Kschool"
type(string)

str

In [ ]:
list = string.split()
print(list)
type(list)

['Python', 'en', 'Kschool']


list

Puedo expecificar el tipo de separador que quiero usar

In [ ]:
list = string.split(" en ")
print(list)
type(list)

['Python', 'Kschool']


list

#### Reemplazar un string
---

La función `replace()` en python es una de las mejores funciones que se pueden aplicar a las cadenas. Por ejemplo, para una cadena **Dinero**, necesitamos reemplazar el signo de dólar y la coma y esto se puede hacer como se muestra a continuación.

In [ ]:
dinero = '$113,678'
print(dinero)
print(type(dinero))

$113,678
<class 'str'>


In [ ]:
dinero = dinero.replace('$', '')
dinero

'113,678'

In [ ]:
dinero = dinero.replace(',', '')
dinero

'113678'

In [ ]:
dinero = int(dinero)
print(dinero)
print(type(dinero))

113678
<class 'int'>


#### Join
---

La función `join()` de python se utiliza para unir la cadena según el patrón especificado.

In [ ]:
string = "Programar en Python es divertido"
string

'Programar en Python es divertido'

In [ ]:
string = " ".join(string)
string

'P r o g r a m a r   e n   P y t h o n   e s   d i v e r t i d o'

#### Capitalize
---

La función `capitalize()` pone en mayúsculas el primer carácter de la palabra o de una cadena.

In [ ]:
string = "alberto"
string

'alberto'

In [ ]:
string = string.capitalize()
string

'Alberto'

#### Center
---

El método `center()` devuelve una cadena rellena con el carácter especificado.

In [ ]:
string = "kschool"
string

'kschool'

In [ ]:
print(string.center(15, '*'))

****kschool****


#### Find
---
El método `find()` devuelve el índice de la posición de la subcadena dada. Si no se encuentra el valor devuelve, -1.

In [ ]:
string = "No tengo el número cuatro"

In [ ]:
print(string.find('ú'))

13


In [ ]:
print(string.find('4'))

-1


#### Strip
---

La función `strip()` elimina los espacios en blanco tanto del inicio como del final de una cadena.

In [ ]:
string = "     la programación no es completa    "
string

'     la programación no es completa    '

In [ ]:
print(string.strip())

la programación no es completa


#### isalnum
---
La función ``isalnum()`` chequea si todos los caracteres de la cadena son alfanuméricos (alfabéticos o numéricos).

In [ ]:
challenge = 'CursodePythonenKscool'
print(challenge.isalnum())

challenge = 'curso de python en kschool'
print(challenge.isalnum())

True
False


### isalpha
---

La función `isalpha` comprueba si todos los elementos de la cadena son caracteres del alfabeto (a-z y A-Z)

In [ ]:
challenge = 'CursodePythonenKscool'
print(challenge.isalpha())
num = '123'
print(num.isalpha())
challenge = 'curso de python en kschool'
print(challenge.isalpha())

True
False
False


#### isdecimal
---

El método `isdecimal()` chequea si todos los caracteres de una cadena son decimales (0-9)

In [ ]:
challenge = 'curso de python en kschool'
print(challenge.isdecimal())
challenge = '123'
print(challenge.isdecimal())

False
True


#### isdigit
---

la función `isdigit()` chequea si todos los caracteres de una cadena son números (0-9 y algunos otros caracteres unicode para números)

In [ ]:
challenge = 'Noventa'
print(challenge.isdigit())
challenge = '90'
print(challenge.isdigit())

False
True


#### isnumeric()
---
La función `isnumeric()` chequea si todos los caracteres de una cadena son números o están relacionados con números (igual que isdigit(), sólo que acepta más símbolos, como ½)

In [ ]:
num = '10'
print(num.isnumeric())
num = '10.5'
print(num.isnumeric())

True
False


#### isidentifier
---
La función `isidentifier()` chequea si para un identificador válido - comprueba si una cadena es un nombre de variable válido

In [ ]:
challenge = '99Curso'
print(challenge.isidentifier())
challenge = 'curso_de_python_en_kschool'
print(challenge.isidentifier())

False
True


#### islower 
---
La función `islower()` chequea si todos los caracteres del alfabeto de la cadena están en minúsculas

In [ ]:
challenge = 'curso de python en kschool'
print(challenge.islower())
challenge = 'curso de python En kschool'
print(challenge.islower())

True
False


#### isupper 
---
La función `isupper()` chequea si todos los caracteres del alfabeto de la cadena están en mayúsculas

In [ ]:
challenge = 'CURSO KSCHOOL'
print(challenge.isupper())
challenge = 'curso de python en kschool'
print(challenge.isupper())

True
False


#### startswith
---

La función `startswith()` chequea si la cadena empieza por la cadena especificada

In [ ]:
challenge = 'curso de python en kschool'
print(challenge.startswith('curso'))

challenge = '90 días de python'
print(challenge.startswith('ninety'))

True
False


Otros métodos: `https://www.programiz.com/python-programming/methods/string`

<h1>¡¡¡Lets Code!!! </h1>
<img src="https://gifdb.com/images/thumbnail/pixel-art-super-mario-computer-amwdq1xi8bgz0omx.gif" alt="Challenge Accepted" style="width: 500px">

#### **Enunciado Ejercicio 1**::
¿Qué longitud tiene la palabra electroencefalografista?

#### **Enunciado Ejercicio 2**::
Por necesidad de mi código necesito que un string dado como input, se recree(repita) tantas veces letras tenga electroencefalografista.

#### **Enunciado Ejercicio 3**::
Dado el siguiente string "Alberto, García García". Necesito separar el nombre del apellido

#### **Enunciado Ejercicio 4**::
¿Me dices en que posición se encuentra el primer número 4 de esta secuencia iukjyhngbfvdcsewgrhtjuykilñkjmnertyutkjn2b5f6dsd4qr536ey7gew645q641245YFQTweqsewdgtrgw43esa?


#### **Enunciado Ejercicio 5**::
¿Usando isnumeric() porque el valor de "3.1416" (PI) es False?

#### **Enunciado Ejercicio 6**::
Ultimamente necesito trabajar en unidades monetarias relacionadas al `€` y no al `$`. Me ayudas a substituir en la siguiente frase el símbolo (no necesito que me hagais el cambio de euros a dólares). "Acabo de volver de visitar a un amigo en Carolina del Norte, ¡Mira el total de dinero que tengo ahora en la cartera es de 126$"


#### **Enunciado Ejercicio 7**::
Se dice que un palíndromo es una palabra o frase que se lee igual en un sentido que en otro. De las siguientes sentencias, ¿cuales son palíndromo?
El bar es imán o zona miserable
Ligar es ser ágil
Aman a Panamá
Ana lleva al oso la avellana